zapcore/field.go

```diff
@@ -65,8 +65,11 @@ const (
 	Int8Type
 	// StringType indicates that the field carries a string.
 	StringType
-	// TimeType indicates that the field carries a time.Time.
+	// TimeType indicates that the field carries a time.Time that is
+	// representable by a UnixNano() stored as an int64.
 	TimeType
+	// TimeFullType indicates that the field carries a time.Time stored as-is.
+	TimeFullType
```

Comment:
Are we sure introducing a new type is okay?
Would it make sense instead to store an invalid value like -1 in the integer field to indicate that the Interface field is set?

Reply:
Since this is defined like an enum, I think it's reasonable to create new values here, but it will break anyone who introspects the internals of a field and only handles the known values.

I don't think we should use the integer field as:
 - The full value range can be used and is valid for existing values. Negative values just mean a time before the epoch. We'd need to truncate the range, but that would cause the following issue.
 - the current `Interface` is storing the timezone, while the new implementation stores a `time.Time` so code that assumes that a timezone is stored for `TimeType` will fail badly (likely panic)

Both options could break callers, but this approach is cleaner, and will only break those who did not assume that the `Type` enum may have new values -- which callers should expect. See similar change in the Go standard library that extends an enum to include a new value:
https://github.com/golang/go/commit/e64241216dd141589144a07f7f68acd64dc108fe

Reply:
Makes sense.